### DISCUSSION TITLE
Midi Keyboard with Velocity & Aftertouch

### DISCUSSION THREAD

**Discussion by RandyRyanG** [tttapa/Control-Surface#932] [Q&A]
I am just learning to code to go easy on me...
I have a 25key midi keyboard with velocity sensitive keys with aftertouch.  I have the keymatrix connected to an arduino micro(32u4). Each key has 2 switches (second switch for velocity) The keymatrix circuit is output to a 16 pin header.

My code is using the NoteButtonmatrix class,  I have added a second button matrix for the "velocity" switches.  I am using MidiView and I am able to see each key press with a clear time delay between the first and second key switch press.  How would I measure the time from the first switch to the second switch to determine the key press velocity? Let me know if I am miss understanding how the 2 switches are intended to work.  This is what my code looks like so far. 

I would also like to connect the built in FSR strip witch I assume is for aftertouch. 

![miniviewsnip](https://github.com/tttapa/Control-Surface/assets/135069048/8fc4afa3-393a-4061-8ee0-d027617c0ea2)

#include <Control_Surface.h>
 
USBMIDI_Interface MPK25V2;

//A0,A2,A4,11 >>>> NOTE ON SWITCHES (addresses)(buttonmatrix)
//A1,A3,A5,12 >>>> VELOCITY ON SWITCHES (addresses2)(buttonmatrix2)
///--------------------------------------------
// The note numbers corresponding to the buttons in the matrix
const AddressMatrix<4, 8> addresses {{
{44, 45, 46, 47, 48, 49, 50, 51},
{52, 53, 54, 55, 56, 57, 58, 59},
{60, 61, 62, 63, 64, 65, 66, 67},
{68, 69, 70, 71, 72, 73, 74, 75},
}};
 

  NoteButtonMatrix<4, 8> buttonmatrix {
  {A0,A2,A4,11}, // row pins
  {2,3,4,5,6,7,8,9},    // column pins
  addresses,    // address matrix
  CHANNEL_1,    // channel and cable number
};
///---------------------------------------------
const AddressMatrix<4, 8> addresses2 {{
{44, 45, 46, 47, 48, 49, 50, 51},
{52, 53, 54, 55, 56, 57, 58, 59},
{60, 61, 62, 63, 64, 65, 66, 67},
{68, 69, 70, 71, 72, 73, 74, 75},
}};
 

  NoteButtonMatrix<4, 8> buttonmatrix2 {
  {A1,A3,A5,12}, // row pins
  {2,3,4,5,6,7,8,9},    // column pins
  addresses2,    // address matrix
  CHANNEL_1,    // channel and cable number
};

void setup() {
  Control_Surface.begin();

}
 
void loop() {
  Control_Surface.loop();

}

**Comment by RandyRyanG** [comment]
@tttapa can you assist please, I know you are busy....

**Reply by RandyRyanG** [reply]
This is what I have so far, everything is working. But I still need some assistance with how I would go about measuring the time interval between button presses of the same key (each key has 2 buttons) to calculate the velocity. I assume I would need to have 2 button matrices one for first button and another for second. Would I still use "notebuttonmatrix" class I know it sends note ons & offs? I would like to get rid of the Filter analog input that is currently setting the velocity.


```
#include <Control_Surface.h>
 
USBMIDI_Interface midi;

// Instantiate a Transposer that can transpose from one octave down to one
// octave up
Transposer<-3, +3> transposer(12);
 
// Instantiate a Selector to change the transposition
IncrementDecrementSelector<transposer.getNumberOfBanks()> selector = {
  transposer,
  {14, 15},
  Wrap::Clamp,
};

// Bank with 10 settings in total, increments of 1 address at a time
//Bank<10> bank {1};

// The note numbers corresponding to the buttons in the matrix
const AddressMatrix<4, 8> addresses {{
{44, 45, 46, 47, 48, 49, 50, 51},
{52, 53, 54, 55, 56, 57, 58, 59},
{60, 61, 62, 63, 64, 65, 66, 67},
{68, 69, 70, 71, 72, 73, 74, 75},
}};

Bankable::NoteButtonMatrix<4, 8> buttonmatrix {
  transposer,  //work fine
  {A1,A3,A5,12}, // row pins
  {2,3,4,5,6,7,8,9}, // column pinss  
  addresses,    // address matrix
  CHANNEL_1,    // channel and cable number
};


// VELOCITY 7-bit filtered analog input for velocity control
FilteredAnalog<7> velocityInput = A10;

 
void setup() {
  Control_Surface.begin();
  buttonmatrix.setVelocity(100);  //0..127 settaggio iniziale velocity
}
 
void loop() {
  if (velocityInput.update())
    buttonmatrix.setVelocity(velocityInput.getValue());
  
  Control_Surface.loop();
}

```

**Reply by RandyRyanG** [reply]
I am still having this issue. Any ideas for calculating the velocity?

**Reply by tttapa** [reply]
Hi, unfortunately, I do not have a lot of free time to play around with this. The existing NoteButtonMatrix class does not support velocity sensitivity. The idea is indeed to record the time when the first switch is pressed (e.g. by saving the `micros()` value into an array), and then when the second switch is activated, you subtract the saved time from thee current time.

To implement this, you'll have to do the scanning of the matrices yourself. The existing implementation can serve as inspiration: https://github.com/tttapa/Control-Surface/blob/main/src/AH/Hardware/ButtonMatrix.ipp#L15-L40

Once you have the time difference between the activation of the two switches, you can map it to a MIDI velocity value from 1 to 127. I assume you can find formulas online or in the literature, and I can imagine it'll take some trial and error to get right.

Finally, you can send the appropriate MIDI note message, see e.g. https://tttapa.github.io/Control-Surface-doc/Doxygen/d3/df7/midi-tutorial.html#midi_md-sending.

**Reply by RandyRyanG** [reply]
Thanks I have this so far, not sure how to "save the micros() value into an array" from the ButtonMatrix I have.
`ButtonMatrix<class Derived, 8, 8> buttonmatrix {
  {25,26,27,28,29,30,31,32}, // row pins odd pins        1,3,5,7,9,11,13,15
  {40,39,38,37,36,35,34,33},    // column pins even pins 2,4,6,8,10,12,14,16

};`

**Reply by tttapa** [reply]
You cannot use the `ButtonMatrix` class like that. You'll either have to create your own replacement of ButtonMatrix, you can re-use some of the code, or you have to create a class that inherits from ButtonMatrix (using [CRTP](https://en.cppreference.com/w/cpp/language/crtp)), see the NoteButtonMatrix class as an example: https://tttapa.github.io/Control-Surface-doc/Doxygen/d7/d33/Abstract_2MIDIButtonMatrix_8hpp_source.html.

**Reply by nds5yf** [reply]
Hi @RandyRyanG,

Have you made any progress implementing Aftertouch either inheriting ButtonMatrix or starting from scratch?
I am working on the same thing, maybe we can collaborate.

Thanks,

Noah
